trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/SmtChcScript.java

```
@@ -0,0 +1,157 @@
+		return mMgdScript.getScript();
+	}
+
+	@Override
+	public LBool solve(final HcSymbolTable symbolTable, final List<HornClause> system) {
+		reset();
+
+		mMgdScript.unlock(this);
+		final var asserter =
+				new ChcAsserter(mMgdScript, getScript(), mProduceUnsatCores, ADD_COMMENTS, DECLARE_FUNCTIONS);
+		asserter.assertClauses(symbolTable, system);
+		mMgdScript.lock(this);
+
+		mName2Clause = asserter.getName2Clause();
+
+		return mMgdScript.checkSat(this);
+	}
+
+	@Override
+	public LBool solve(final HcSymbolTable symbolTable, final List<HornClause> system, final long timeout) {
+		mMgdScript.getScript().setOption(":timeout", timeout);
+		try {
+			return solve(symbolTable, system);
+		} finally {
+			try {
+				// reset timeout
+				mMgdScript.getScript().setOption(":timeout", 0L);
+			} catch (final Throwable e) {
+				// swallow exception
+				e.printStackTrace();
+			}
+		}
+	}
+
+	@Override
+	public boolean supportsModelProduction() {
+		return true;
+	}
+
+	@Override
+	public void produceModels(final boolean enable) {
+		getScript().setOption(SMTLIBConstants.PRODUCE_MODELS, enable);
+	}
+
+	@Override
+	public Optional<Model> getModel() {
+		return Optional.ofNullable(getScript().getModel());
+	}
+
+	@Override
+	public boolean supportsDerivationProduction() {
+		return false;
+	}
+
+	@Override
+	public void produceDerivations(final boolean enable) {
+		throw new UnsupportedOperationException("Derivations are not supported");
+	}
+
+	@Override
+	public Optional<Derivation> getDerivation() {
+		throw new UnsupportedOperationException("Derivations are not supported");
+	}
+
+	@Override
+	public boolean supportsUnsatCores() {
+		return true;
+	}
+
+	@Override
+	public void produceUnsatCores(final boolean enable) {
+		getScript().setOption(SMTLIBConstants.PRODUCE_UNSAT_CORES, enable);
+		mProduceUnsatCores = enable;
+	}
+
+	@Override
+	public Optional<Set<HornClause>> getUnsatCore() {
+		final var core = mMgdScript.getUnsatCore(this);
```

dont you want to check mProduceUnsatCores?

---

I'm unsure. You mean we should throw an exception if it's `false` ?

Technically the behaviour of this method is unspecified if UNSAT cores were not previously enabled. It might be a good idea to catch such errors early though... But then I should probably check that consistently, i.e., also for `getModel()`, and similarly in `GolemChcScript`.